packages/specs/source/01-app-definition.md

```diff
@@ -1,40 +1,169 @@
 
-From certain app states, `resolve` can be called to return a value of type `struct Transfer.Transaction` (this is defined by framework code in `Transfer.sol`). This allows the state deposit assigned to the app to be reassigned, for e.g., to the winner of the Tic-Tac-Toe game.
+When writing a state channel application presently, we require that the resolution of an application be directly tied one-to-one to some particular state that is being progressed. For example, when defining a game of Tic-Tac-Toe, we ask that the resolution function which checks for the winner specifically return a data structure that can be interpretted for sending ETH to the user that won.
+
+Here is a diagram that shows how a resolution looks presently for a game of Tic-Tac-Toe:
 
 ![](img/resolve.svg)
 
-### isStateTerminal
+**Note:** This is subject to an improvement that will be added soon that will de-couple an application from the exact transaction that occurs upon resolution to an interpreter pattern that will allow for multiple assets to be re-used on the same state machine logic. For example, a Tic-Tac-Toe game either returns a full-amount for one person or divides an amount in half; those are the only two options. These two options can be mapped to ETH, ERC20 tokens, or other assets which are fungible in single or half quantities.
+
+## AppDefinition
+
+To address all of the above requirements of state channel applications, we introduce an interface called an `AppDefinition` which **implements the logic of an application in the EVM**. To be specific, the `AppDefinition` interface is one which can be implemented by a developer that is interested in writing a state channels based application that the Counterfactual project supports in the rest of the framework (e.g., in the [adjudication layer](./02-adjudication-layer.md)).
+
+### Resolution Function
+
+There is a single function which _must_ be implemented in the interface. This function is essentially providing the resolution functionality discussed above. The function signature looks like this:
+
+```solidity
+function resolve(bytes memory, Transfer.Terms memory) public view returns (Transfer.Transaction);
+```
+
+The first argument which is of type `bytes memory` is an internally-referencable state object for the `AppDefinition`. For example, in the case of a payment channel it must be considered as the ABI-encoded version of a structure encoding the type from above. This means that you will likely want to use `abi.decode` inside of the `resolve` method to decode the bytes into a usable struct object. In our payment channel example this would look like:
```

Comment:
```suggestion
The first argument of type `bytes memory` is an internally-referencable state object for the `AppDefinition`. For example, in the case of a payment channel it must be considered as the ABI-encoded version of a structure encoding the type from above. This means that you will likely want to use `abi.decode` inside of the `resolve` method to decode the bytes into a usable struct object. In our payment channel example this would look like:
```